pymc3/distributions/distribution.py

```diff
@@ -375,110 +392,94 @@ def generate_samples(generator, *args, **kwargs):
-
-    Parameters
-    ----------
-    distribution : pymc3 distribution
-        Distribution to be transformed into a bounded distribution
-    lower : float (optional)
-        Lower bound of the distribution, set to -inf to disable.
-    upper : float (optional)
-        Upper bound of the distribibution, set to inf to disable.
-    tranform : 'infer' or object
-        If 'infer', infers the right transform to apply from the supplied bounds.
-        If transform object, has to supply .forward() and .backward() methods.
-        See pymc3.distributions.transforms for more information.
-    """
-
-    def __init__(self, distribution, lower, upper, transform='infer', *args, **kwargs):
-        import pymc3.distributions.transforms as transforms
-        self.dist = distribution.dist(*args, **kwargs)
-
-        self.__dict__.update(self.dist.__dict__)
-        self.__dict__.update(locals())
-
-        if hasattr(self.dist, 'mode'):
-            self.mode = self.dist.mode
-
-        if transform == 'infer':
-
-            default = self.dist.default()
-
-            if not np.isinf(lower) and not np.isinf(upper):
-                self.transform = transforms.interval(lower, upper)
-                if default <= lower or default >= upper:
-                    self.testval = 0.5 * (upper + lower)
-
-            if not np.isinf(lower) and np.isinf(upper):
-                self.transform = transforms.lowerbound(lower)
-                if default <= lower:
-                    self.testval = lower + 1
-
-            if np.isinf(lower) and not np.isinf(upper):
-                self.transform = transforms.upperbound(upper)
-                if default >= upper:
-                    self.testval = upper - 1
-
-        if issubclass(distribution, Discrete):
-            self.transform = None
-
-    def _random(self, lower, upper, point=None, size=None):
-        samples = np.zeros(size).flatten()
-        i, n = 0, len(samples)
-        while i < len(samples):
-            sample = self.dist.random(point=point, size=n)
-            select = sample[np.logical_and(sample > lower, sample <= upper)]
-            samples[i:(i + len(select))] = select[:]
-            i += len(select)
-            n -= len(select)
-        if size is not None:
-            return np.reshape(samples, size)
-        else:
-            return samples
-
-    def random(self, point=None, size=None, repeat=None):
-        lower, upper = draw_values([self.lower, self.upper], point=point)
-        return generate_samples(self._random, lower, upper, point,
-                                dist_shape=self.shape,
-                                size=size)
-
-    def logp(self, value):
-        return bound(self.dist.logp(value),
-                     value >= self.lower, value <= self.upper)
+class _BoundedIndicator(object):
+    pass
 
 
-class Bound(object):
+def Bounded(distribution, lower=-np.inf, upper=np.inf, transform='infer'):
```

Comment:
Seems quite costly in terms of added code complexity. I suppose `isinstance(Bound(), Distribution)` returns True?